### ISSUE TITLE
Questions about the transferlearning experiments

### ISSUE THREAD

**Issue by ha-lins** [Shen-Lab/GraphCL#21]
Hi @yyou1996,

I tired pretraining on transferlearning experiments yet found: 1) on Bio benchmark, the loss and acc are always 0.0 and nan; 2) On chem dataset, the training process stopped at the 2nd epoch. Could you pls check the code? Btw, the finetuning command worked well.  Thanks a lot!

![image](https://user-images.githubusercontent.com/22448743/114803273-c09a5e80-9dd1-11eb-84f1-6e1731f0a761.png)

![image](https://user-images.githubusercontent.com/22448743/114803228-af515200-9dd1-11eb-827d-a061cc2bdfb2.png)

Ps: My environment is `cuda=10.0` and pyG has the same version just except it is for `cu100`.

**Comment by yyou1996** [comment]
Hi @ha-lins,

The correct command to run is `python pretrain_graphcl.py --aug1 random --aug2 random`, otherwise the default aug is none.

And for the acc part, since pretraining does not contribute to supervised acc, I just let acc=0 and print. So just ignore it.
https://github.com/Shen-Lab/GraphCL/blob/7eefcc3ca3e0c9a579fd17bcb06fd28df9733312/transferLearning_MoleculeNet_PPI/bio/pretrain_graphcl.py#L103

**Comment by ha-lins** [comment]
Hi @yyou1996,

I met the same problem if I set the `aug.` parameter and the training process stopped at the second epoch. Could you give some suggestions? Thanks!
 
![image](https://user-images.githubusercontent.com/22448743/130039672-75984361-b0d8-474f-ad90-3793443b4fc9.png)

**Comment by yyou1996** [comment]
Hi @ha-lins,

Is it also the case if you set both as `random`?

**Comment by ha-lins** [comment]
Yes, it's the same case if I set both as `random` @yyou1996 :
![image](https://user-images.githubusercontent.com/22448743/130115103-42ce8721-502b-4261-bc81-4cd9e98e3cd4.png)

It's puzzling since the first epoch runs well while the second epoch not.

**Comment by ha-lins** [comment]
Yes, it's the same case if I set both as `random` @yyou1996 :
![image](https://user-images.githubusercontent.com/22448743/130115103-42ce8721-502b-4261-bc81-4cd9e98e3cd4.png)

It's puzzling since the first epoch runs well while the second epoch not.

**Comment by yyou1996** [comment]
@ha-lins 

I suspect might be the memory issue. Would you try adding `del dataset1` and `del dataset 2` at the end of this `train()` function (before return)? Also monitor the memory (e.g. use `htop`) to see whether it is the problem.
https://github.com/Shen-Lab/GraphCL/blob/d857849d51bb168568267e07007c0b0c8bb6d869/transferLearning_MoleculeNet_PPI/chem/pretrain_graphcl.py#L75

**Comment by ha-lins** [comment]
@yyou1996 ,

Thanks for your solution. I tried it but it didn't work. With `htop`, I found that the main process and all sub-process `missed in the second epoch`. And I ended the job manually and found it always stopped at the same place of `shuffle` and [get()](https://github.com/Shen-Lab/GraphCL/blob/master/transferLearning_MoleculeNet_PPI/chem/loader.py#L302):

![image](https://user-images.githubusercontent.com/22448743/130178866-e3e0c62d-1959-4e0f-8fb5-bcc700669778.png)

Is there any other ideas? Thanks!

**Comment by yyou1996** [comment]
@ha-lins 

Looks like the process in your machine is somehow killed. I just check their [`shuffle`](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/data/dataset.html#Dataset.shuffle) function and they do have `copy` applied. So my suspect might be: in epoch 1 memory is under limit, everything work ok. When in epoch 2, with in `shuffle` there is `copy` that double(?) the memory, leading to the process killed.

This might be solved if at the end of epoch 1, all memory of dataset and dataloader are clear. So I would try `del dataset1, del dataset2, del dataloader1, del dataloader2`. And according to this [link](https://www.delftstack.com/howto/python/python-clear-memory/) probably `gc.collect()` is also needed after.

Last last, FYI the new version https://github.com/Shen-Lab/GraphCL_Automated/tree/master/transferLearning_MoleculeNet_PPI do not apply `copy` function anymore so it shouldn't have the above problem.

**Comment by ha-lins** [comment]
@yyou1996 

`gc.collect()` is somehow helpful. But I have to kill the job manually (`ctrl+c`), then it can go on. If not, it still faces the stopping problem. So I have to kill for each epoch. Is there any automatic quit way?

![image](https://user-images.githubusercontent.com/22448743/130624965-79f4bf6e-9754-4b70-b59f-e41ac678d877.png)

The new version of GraphCL_Automated is good while it needs too much revision. So I would try the first solution further.

**Comment by yyou1996** [comment]
@ha-lins 

Looks like dataloader is stuck in multiple process. Try to set `num_workers=1` for dataloader? (e.g. https://github.com/Shen-Lab/GraphCL/blob/d857849d51bb168568267e07007c0b0c8bb6d869/transferLearning_MoleculeNet_PPI/chem/pretrain_graphcl.py#L83)

**Comment by yyou1996** [comment]
Hi @ha-lins,

Does it solve the problem? For faster feedback recycle maybe you can email me for discussion. I will post our solution in this issue lastly.

**Comment by ha-lins** [comment]
@yyou1996,
 
The problem was not solved by set `num_workers = 1 or 0`. I change to another GPU machine, rtx 3090 with this [environment](https://github.com/Shen-Lab/GraphCL/issues/25#issuecomment-843048686) and it works now. 
Anyway, thanks for your continual help. : )
